{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing through JSON files to obtain dataframe with mutations found on Nextstrain's visualization of H7N9 IAV\n",
    "\n",
    "Author: Luis Antonio Haddock III & Katarina M Braun, MD/PhD \n",
    "\n",
    "Adapted from: Louise Moncla, PhD (https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1008191)\n",
    "\n",
    "\n",
    "Last edit: 2022-03-27\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, subprocess, glob, os, shutil, re, importlib, csv, json\n",
    "from subprocess import call\n",
    "import collections\n",
    "from collections import Counter\n",
    "from Bio import SeqIO\n",
    "from Bio import Seq\n",
    "import Bio.Phylo\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import rpy2\n",
    "from scipy import stats\n",
    "%reload_ext rpy2.ipython\n",
    "from io import BytesIO as csio\n",
    "import requests\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "from functools import reduce\n",
    "\n",
    "# we might not use all modules in this analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "segments = ['ha', 'pa', 'na', 'pb1', 'pb2', 'mp', 'ns', 'np']\n",
    "\n",
    "\n",
    "for file in segments:\n",
    "    links = ['http://data.nextstrain.org/flu_avian_h7n9_' + file + '.json']\n",
    "    \n",
    "    for link in links:\n",
    "        name = link.split(\"/\")[3].lower()\n",
    "        path = ['JSON_files/' + name]\n",
    "        tree_json=json.load(csio(requests.get(link).content))\n",
    "        tree_json_2 = tree_json['tree']\n",
    "        out_file = open(name, \"w\")\n",
    "        json.dump(tree_json_2, out_file, indent = 2) \n",
    "        out_file.close() \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# creating a directory to store all json files (internal use only)\n",
    "\n",
    "rm -rf JSON_files/\n",
    "mkdir -p JSON_files/\n",
    "\n",
    "mv *.json JSON_files/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to read json file and store it in a variable \n",
    "\n",
    "def read_json(file_name):\n",
    "    try:\n",
    "        handle = open(file_name, 'r')\n",
    "    except IOError:\n",
    "        pass\n",
    "    else:\n",
    "        data = json.load(handle)\n",
    "        handle.close()\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reading the tree json file in V2 (different code is used for old V1 JSON format)\n",
    "\n",
    "def json_to_tree(json_dict, root=True):\n",
    "    # Check for v2 JSON which has combined metadata and tree data.\n",
    "    \n",
    "    if root and \"meta\" in json_dict and \"tree\" in json_dict:\n",
    "        json_dict = json_dict[\"tree\"]\n",
    "    \n",
    "    node = Bio.Phylo.Newick.Clade()\n",
    "    \n",
    "    if \"name\" in json_dict:\n",
    "        node.name = json_dict[\"name\"] #strain was changed to name from v1 to v2\n",
    "    else:\n",
    "        node.name = json_dict[\"strain\"]\n",
    "\n",
    "    if \"children\" in json_dict:\n",
    "        # Recursively add children to the current node.\n",
    "        node.clades = [json_to_tree(child, root=False) for child in json_dict[\"children\"]]\n",
    "\n",
    "    # Assign all non-children attributes.\n",
    "    for attr, value in json_dict.items():\n",
    "        if attr != \"children\":\n",
    "            setattr(node, attr, value)\n",
    "\n",
    "   # Only v1 JSONs support a single `attr` attribute.\n",
    "    if hasattr(node, \"attr\"):\n",
    "        node.numdate = node.attr.get(\"num_date\")\n",
    "        node.branch_length = node.attr.get(\"div\")\n",
    "\n",
    "        if \"translations\" in node.attr:\n",
    "            node.translations = node.attr[\"translations\"]\n",
    "    elif hasattr(node, \"node_attrs\"):\n",
    "        node.branch_length = node.node_attrs.get(\"div\")\n",
    "\n",
    "    if root:\n",
    "        node = annotate_parents(node)\n",
    "\n",
    "        \n",
    "    return node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotations \n",
    "\n",
    "def annotate_parents(tree):\n",
    "    # Get all parent nodes by node.\n",
    "    parents_by_node = all_parents(tree)\n",
    "    #print(\"number of parent nodes is: \", len(parents_by_node))\n",
    "\n",
    "    # Next, annotate each node with its parent.\n",
    "    for node in tree.find_clades():\n",
    "        if node == tree.root:\n",
    "            node.up = None\n",
    "        else:\n",
    "            node.up = parents_by_node[node]\n",
    "\n",
    "    # Return the tree.\n",
    "    return tree\n",
    "\n",
    "def all_parents(tree):\n",
    "    parents = {}\n",
    "    for clade in tree.find_clades(order='level'):\n",
    "        for child in clade:\n",
    "            parents[child] = clade\n",
    "    return parents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parsing thorugh tree in V2 (adapted from Louise Moncla, PhD)\n",
    "\n",
    "def parse_tree_and_return_muts(tree, gene): \n",
    "    \n",
    "\n",
    "#lists to store aa mutations\n",
    "    avian_mutations = []\n",
    "    human_mutations = []\n",
    "    \n",
    "    for clade in tree.find_clades(): \n",
    "        tips_list = []\n",
    "        hosts_list = []\n",
    "        host_values = []\n",
    "        \n",
    "        for terminal in clade.get_terminals():\n",
    "            tips_list.append(terminal.name)\n",
    "            hosts_list.append(terminal.node_attrs['host'])\n",
    "            \n",
    "        for a_dict in hosts_list:\n",
    "            host_values.append(a_dict['value'])\n",
    "            \n",
    "            if set(host_values) == {\"Human\"}:\n",
    "                \n",
    "                up1_hosts_list = []\n",
    "                up1_temp_host = []\n",
    "                up1_tips_list = []\n",
    "                \n",
    "                if clade.up == None:\n",
    "                    print(\"root\")\n",
    "                \n",
    "                else:\n",
    "                    for x in clade.up.get_terminals():\n",
    "                        up1_tips_list.append(x.name)\n",
    "                        up1_temp_host.append(x.node_attrs['host'])\n",
    "                    \n",
    "\n",
    "                    for b_dict in up1_temp_host:\n",
    "                        up1_hosts_list.append(b_dict['value'])\n",
    "\n",
    "                    #we look through nodes for human hosts and extract the \n",
    "                    #mutations that leads to said branch\n",
    "                    if set(up1_hosts_list) == {'Human'} and 'mutations' in clade.branch_attrs and 'mutations' in clade.branch_attrs != {}:  \n",
    "                        if gene in clade.branch_attrs['mutations']:\n",
    "\n",
    "                            aa_muts = clade.branch_attrs['mutations']\n",
    "                            \n",
    "                            for mut in aa_muts[gene]:\n",
    "\n",
    "                                human_mutations.append(mut)\n",
    "        \n",
    "                            \n",
    "            \n",
    "            \n",
    "        for a_dict in hosts_list:\n",
    "            host_values.append(a_dict['value'])\n",
    "            \n",
    "            if set(host_values) == {\"Avian\"}:\n",
    "                \n",
    "                up2_hosts_list = []\n",
    "                up2_temp_host = []\n",
    "                up2_tips_list = []\n",
    "                \n",
    "                #for x in clade.get_terminals(): #if I fo this then it works with HA\n",
    "                \"\"\"add same thing down here\"\"\"\n",
    "                if clade.up == None: \n",
    "                    print(\"root\")\n",
    "                else:\n",
    "                    for x in clade.up.get_terminals():\n",
    "                        up2_tips_list.append(x.name)\n",
    "                        up2_temp_host.append(x.node_attrs['host'])\n",
    "\n",
    "                    for b_dict in up2_temp_host:\n",
    "                        up2_hosts_list.append(b_dict['value'])\n",
    "\n",
    "                    # if the node above contains avan samples, then clade should be set to avian\n",
    "                    if set(up2_hosts_list) == {'Avian'} and 'mutations' in clade.branch_attrs and 'mutations' in clade.branch_attrs != {}:  \n",
    "                        if gene in clade.branch_attrs['mutations']:\n",
    "\n",
    "                            aa_muts = clade.branch_attrs['mutations']\n",
    "                            for mut in aa_muts[gene]:\n",
    "\n",
    "                                avian_mutations.append(mut)\n",
    "\n",
    "    \n",
    "    return(avian_mutations, human_mutations)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Counting the number of times each mutation occurrs in each list\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For simplicity we change aa code format\n",
    "\n",
    "aa_dict = {\"A\":\"Ala\",\"R\":\"Arg\",\"N\":\"Asn\",\"D\":\"Asp\",\"C\":\"Cys\",\"Q\":\"Gln\",\"E\":\"Glu\",\"G\":\"Gly\",\"H\":\"His\",\n",
    "           \"I\":\"Ile\",\"L\":\"Leu\",\"K\":\"Lys\",\"M\":\"Met\",\"F\":\"Phe\",\"P\":\"Pro\",\"S\":\"Ser\",\"T\":\"Thr\",\"W\":\"Trp\",\n",
    "           \"Y\":\"Tyr\",\"V\":\"Val\", \"*\":\"stop\"}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_aa_changes(bb, hh):\n",
    "    bb2 = []\n",
    "    #bh2 = []\n",
    "    hh2 = []\n",
    "    \n",
    "    for a in bb:\n",
    "        # convert aa symbols to 3-letter abbreviations\n",
    "        first_aa = a[0]\n",
    "        new_first = aa_dict[first_aa]\n",
    "        last_aa = a[-1]\n",
    "        new_last = aa_dict[last_aa]\n",
    "        a = a.replace(first_aa, new_first)\n",
    "        a = a.replace(last_aa, new_last)\n",
    "        bb2.append(a)  \n",
    "#    for a in bh:\n",
    "#        # convert aa symbols to 3-letter abbreviations\n",
    "#        first_aa = a[0]\n",
    "#        new_first = aa_dict[first_aa]\n",
    "#        last_aa = a[-1]\n",
    "#        new_last = aa_dict[last_aa]\n",
    "#        a = a.replace(first_aa, new_first)\n",
    "#        a = a.replace(last_aa, new_last)\n",
    "#        bh2.append(a)\n",
    "    for a in hh:\n",
    "        # convert aa symbols to 3-letter abbreviations\n",
    "        first_aa = a[0]\n",
    "        new_first = aa_dict[first_aa]\n",
    "        last_aa = a[-1]\n",
    "        new_last = aa_dict[last_aa]\n",
    "        a = a.replace(first_aa, new_first)\n",
    "        a = a.replace(last_aa, new_last)\n",
    "        hh2.append(a)\n",
    "        \n",
    "    return(bb2, bh2, hh2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_SNP_counts(bb, hh):\n",
    "    \n",
    "    # use Counter to count the number of times each amino acid change is detected in each list; print total unique SNPs\n",
    "    #bird_to_bird_count = Counter(bird_to_bird)\n",
    "    bird = Counter(avian_mutations)\n",
    "    human = Counter(human_mutations)\n",
    "    \n",
    "    # get a complete list of all of the SNPs identified\n",
    "    all_SNPs = set(avian_mutations + human_mutations)\n",
    "    \n",
    "    # loop through and count how many times each SNP occurs in each dataset \n",
    "    all_counts = {}\n",
    "\n",
    "    for a in all_SNPs:\n",
    "                \n",
    "        if a in bird:\n",
    "            bird_snp = bird[a]\n",
    "        else:\n",
    "            bird_snp = 0\n",
    "\n",
    "        if a in human:\n",
    "            human_snp = human[a]   \n",
    "        else:\n",
    "            human_snp = 0\n",
    "\n",
    "\n",
    "        all_counts[a] = {\"bird_snp\":bird_snp, \"human_snp\": human_snp}\n",
    "    return(all_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_dataframe(all_counts, gene, df):\n",
    "    df1 = pd.DataFrame.from_dict(all_counts, orient='index')\n",
    "    df1 = df1.reset_index()\n",
    "    df1.columns=['coding_region_change','avian_mutations','human_mutations'] #, 'avian_strains', 'human_strains']\n",
    "    df1['gene'] = gene\n",
    "    df = pd.concat([df,df1])\n",
    "    return(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running functions across JSON files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      "root\n",
      "root\n",
      "root\n",
      "root\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Length mismatch: Expected axis has 1 elements, new values have 3 elements",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [17]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     29\u001b[0m \u001b[38;5;66;03m#bird_to_bird, bird_to_human, human_to_human = convert_aa_changes(bird_to_bird, bird_to_human, human_to_human)\u001b[39;00m\n\u001b[1;32m     30\u001b[0m all_counts \u001b[38;5;241m=\u001b[39m return_SNP_counts(avian_mutations, human_mutations)\n\u001b[0;32m---> 31\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[43mgenerate_dataframe\u001b[49m\u001b[43m(\u001b[49m\u001b[43mall_counts\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdf\u001b[49m\u001b[43m)\u001b[49m\n",
      "Input \u001b[0;32mIn [16]\u001b[0m, in \u001b[0;36mgenerate_dataframe\u001b[0;34m(all_counts, gene, df)\u001b[0m\n\u001b[1;32m      2\u001b[0m df1 \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame\u001b[38;5;241m.\u001b[39mfrom_dict(all_counts, orient\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mindex\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      3\u001b[0m df1 \u001b[38;5;241m=\u001b[39m df1\u001b[38;5;241m.\u001b[39mreset_index()\n\u001b[0;32m----> 4\u001b[0m df1\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcoding_region_change\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mavian_mutations\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhuman_mutations\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;66;03m#, 'avian_strains', 'human_strains']\u001b[39;00m\n\u001b[1;32m      5\u001b[0m df1[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgene\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m gene\n\u001b[1;32m      6\u001b[0m df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat([df,df1])\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pandas/core/generic.py:5596\u001b[0m, in \u001b[0;36mNDFrame.__setattr__\u001b[0;34m(self, name, value)\u001b[0m\n\u001b[1;32m   5594\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   5595\u001b[0m     \u001b[38;5;28mobject\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__getattribute__\u001b[39m(\u001b[38;5;28mself\u001b[39m, name)\n\u001b[0;32m-> 5596\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mobject\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__setattr__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   5597\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mAttributeError\u001b[39;00m:\n\u001b[1;32m   5598\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pandas/_libs/properties.pyx:70\u001b[0m, in \u001b[0;36mpandas._libs.properties.AxisProperty.__set__\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pandas/core/generic.py:769\u001b[0m, in \u001b[0;36mNDFrame._set_axis\u001b[0;34m(self, axis, labels)\u001b[0m\n\u001b[1;32m    767\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_set_axis\u001b[39m(\u001b[38;5;28mself\u001b[39m, axis: \u001b[38;5;28mint\u001b[39m, labels: Index) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    768\u001b[0m     labels \u001b[38;5;241m=\u001b[39m ensure_index(labels)\n\u001b[0;32m--> 769\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_mgr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mset_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlabels\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    770\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_clear_item_cache()\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pandas/core/internals/managers.py:214\u001b[0m, in \u001b[0;36mBaseBlockManager.set_axis\u001b[0;34m(self, axis, new_labels)\u001b[0m\n\u001b[1;32m    212\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mset_axis\u001b[39m(\u001b[38;5;28mself\u001b[39m, axis: \u001b[38;5;28mint\u001b[39m, new_labels: Index) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    213\u001b[0m     \u001b[38;5;66;03m# Caller is responsible for ensuring we have an Index object.\u001b[39;00m\n\u001b[0;32m--> 214\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_set_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnew_labels\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    215\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxes[axis] \u001b[38;5;241m=\u001b[39m new_labels\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pandas/core/internals/base.py:69\u001b[0m, in \u001b[0;36mDataManager._validate_set_axis\u001b[0;34m(self, axis, new_labels)\u001b[0m\n\u001b[1;32m     66\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n\u001b[1;32m     68\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m new_len \u001b[38;5;241m!=\u001b[39m old_len:\n\u001b[0;32m---> 69\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m     70\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLength mismatch: Expected axis has \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mold_len\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m elements, new \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     71\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvalues have \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnew_len\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m elements\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     72\u001b[0m     )\n",
      "\u001b[0;31mValueError\u001b[0m: Length mismatch: Expected axis has 1 elements, new values have 3 elements"
     ]
    }
   ],
   "source": [
    "#old_code\n",
    "df = pd.DataFrame()\n",
    "\n",
    "for f in glob.glob(\"JSON_files/flu_avian_h7n9_*.json\"):\n",
    "    tree = read_json(f)\n",
    "    tree = json_to_tree(tree)\n",
    "    gene = f.split(\"_\")[3].upper()\n",
    "    gene = gene.split(\".\")[0].upper()\n",
    "\n",
    "    \n",
    "    if gene == \"MP\":\n",
    "        genes = [\"M1\",\"M2\"]\n",
    "        for g in genes: \n",
    "            avian_mutations, human_mutations  = parse_tree_and_return_muts(tree, g)\n",
    "#            #bird_to_bird, bird_to_human, human_to_human = convert_aa_changes(bird_to_bird, bird_to_human, human_to_human)\n",
    "            all_counts = return_SNP_counts(avian_mutations, human_mutations )\n",
    "            df = generate_dataframe(all_counts, g, df)\n",
    "    \n",
    "    elif gene == \"NS\":\n",
    "        genes = [\"NS1\",\"NS2\"]\n",
    "        for g in genes: \n",
    "            avian_mutations, human_mutations  = parse_tree_and_return_muts(tree, g)\n",
    "            #bird_to_bird, bird_to_human, human_to_human = convert_aa_changes(bird_to_bird, bird_to_human, human_to_human)\n",
    "            all_counts = return_SNP_counts(avian_mutations, human_mutations )\n",
    "            df = generate_dataframe(all_counts, g, df)\n",
    "    elif gene != \"MP\" and gene != \"NS\": # and gene != 'HA': \n",
    "        # run everything\n",
    "        avian_mutations, human_mutations = parse_tree_and_return_muts(tree, gene)\n",
    "        #bird_to_bird, bird_to_human, human_to_human = convert_aa_changes(bird_to_bird, bird_to_human, human_to_human)\n",
    "        all_counts = return_SNP_counts(avian_mutations, human_mutations)\n",
    "        df = generate_dataframe(all_counts, gene, df)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([], dtype=object)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df.to_csv('mutations_with_strains.tsv', sep='\\t')\n",
    "\n",
    "df['gene'] = df['gene'].str.replace('NS1', 'NS')\n",
    "df['gene'] = df['gene'].str.replace('NS2', 'NS')\n",
    "df['gene'] = df['gene'].str.replace('M1', 'MP')\n",
    "df['gene'] = df['gene'].str.replace('M2', 'MP')\n",
    "df = df.drop_duplicates()\n",
    "df['gene'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coding_region_change</th>\n",
       "      <th>avian_mutations</th>\n",
       "      <th>human_mutations</th>\n",
       "      <th>gene</th>\n",
       "      <th>region_gene</th>\n",
       "      <th>site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q400K</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>HA</td>\n",
       "      <td>Q400K_HA</td>\n",
       "      <td>HA Q400K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L517I</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>HA</td>\n",
       "      <td>L517I_HA</td>\n",
       "      <td>HA L517I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A169T</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>HA</td>\n",
       "      <td>A169T_HA</td>\n",
       "      <td>HA A169T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M547I</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>HA</td>\n",
       "      <td>M547I_HA</td>\n",
       "      <td>HA M547I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>D283A</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>HA</td>\n",
       "      <td>D283A_HA</td>\n",
       "      <td>HA D283A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>L136M</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NP</td>\n",
       "      <td>L136M_NP</td>\n",
       "      <td>NP L136M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>F420Y</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NP</td>\n",
       "      <td>F420Y_NP</td>\n",
       "      <td>NP F420Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>T215A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NP</td>\n",
       "      <td>T215A_NP</td>\n",
       "      <td>NP T215A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>T442A</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NP</td>\n",
       "      <td>T442A_NP</td>\n",
       "      <td>NP T442A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>E127D</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NP</td>\n",
       "      <td>E127D_NP</td>\n",
       "      <td>NP E127D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2071 rows √ó 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    coding_region_change  avian_mutations  human_mutations gene region_gene  \\\n",
       "0                  Q400K                0                1   HA    Q400K_HA   \n",
       "1                  L517I                8                0   HA    L517I_HA   \n",
       "2                  A169T                2                7   HA    A169T_HA   \n",
       "3                  M547I                0                1   HA    M547I_HA   \n",
       "4                  D283A                0                5   HA    D283A_HA   \n",
       "..                   ...              ...              ...  ...         ...   \n",
       "215                L136M                0                1   NP    L136M_NP   \n",
       "216                F420Y                0                1   NP    F420Y_NP   \n",
       "217                T215A                0                1   NP    T215A_NP   \n",
       "218                T442A                1                0   NP    T442A_NP   \n",
       "219                E127D                0                1   NP    E127D_NP   \n",
       "\n",
       "         site  \n",
       "0    HA Q400K  \n",
       "1    HA L517I  \n",
       "2    HA A169T  \n",
       "3    HA M547I  \n",
       "4    HA D283A  \n",
       "..        ...  \n",
       "215  NP L136M  \n",
       "216  NP F420Y  \n",
       "217  NP T215A  \n",
       "218  NP T442A  \n",
       "219  NP E127D  \n",
       "\n",
       "[2071 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['region_gene'] = df['coding_region_change'] + \"_\" + df['gene']\n",
    "df['site'] = df['gene'] + \" \" + df['coding_region_change']\n",
    "df.to_csv('full_nextstrain_mutations_table.tsv', sep='\\t')\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Success, your code works!!! üôèüèª\n"
     ]
    }
   ],
   "source": [
    "print('Success, your code works!!! üôèüèª')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
